Mjx

Vittorio-Caggiano opened the pull request:
## Summary
This PR adds/expands the **MJX (JAX/MuJoCo-JAX)** integration for MyoSuite, including new MJX environments, training/benchmark scripts, and MJX-focused unit tests. It also modernizes dependency management and CI/docs by switching workflows to **`uv`**, adding/updating **`uv.lock`**, and tightening Python/ML dependency constraints.

## Key changes
- **MJX backend**
  - New `myosuite/envs/myo/mjx/` package (MJX env implementation, registry, utilities, training/visualization/playground helpers).
  - JAX implementations for core components (e.g., quaternion math) plus MJX-specific fatigue/reference motion handling.
- **Benchmarks & results**
  - Added multiple MJX benchmark scripts under `benchmarks/` (training + PPO variants + baseline + shell entrypoints).
  - Updated MJX README and added benchmark figures/results assets.
- **Tests**
  - Added/extended MJX unit tests under `myosuite/tests/mjx/` (fatigue, quat math, reference motion).
  - Updated test runners to cover MJX components where applicable.
- **Dependency / packaging updates**
  - Updated `pyproject.toml`:
    - `requires-python` -> `>=3.10,<=3.12`
    - Relaxed/ranged key dependencies (e.g., `gymnasium`, `mujoco`)
    - Added optional dependency groups: `mjx`, `mjx-cuda`, `docs`, and `dependency-groups.dev`
    - Added `uv` index configuration for `mujoco` and `nvidia` packages
  - Added/updated `uv.lock`.
- **CI / release / docs**
  - Updated GitHub workflows to use `uv` for installs/tests.
  - Added a docs build workflow (`docs.yml`) using `uv` + Sphinx.
  - Release workflow updated to also update `uv.lock` when bumping versions.
  - ReadTheDocs config updated (OS/Python versions).
- **Misc maintenance**
  - `.gitignore` updates, submodule URL update, and broader docs/tutorial refresh.

## Testing
- CI runs MJX tests via `pytest` using the updated `uv`-based workflow and `.[mjx]` extra.
- Local verification suggestion (mirrors CI):
  - `uv run --with ".[mjx]" python -m pytest ...`

jamesheald committed:
add mjx folder

Balint-H committed:
Add mujoco-playground style elbow environment

Balint-H committed:
Add mujoco-playground style elbow environment

Balint-H committed:
Fix typos

Balint-H committed:
Add interactive elbow eval script

Balint-H committed:
Remove unnecessary bias to elbow reset angle

Balint-H committed:
Remove leftover variables from old brax environment

jamesheald committed:
initial ial progress on generalizing elbow environment to pose; elbow model programatically modified with mjspec

jamesheald committed:
Remove accidentally committed log files

Balint-H committed:
Refine visualization environments and use base class in elbow

Balint-H committed:
Add hand example environment also using the pose base

elladyr committed:
Restore head site in chasetag

jamesheald committed:
reach base class created

jamesheald committed:
remove time from obs; set model variables before calling put_model; normalize actions using sigmoid

jamesheald committed:
explicitly set solver parameters during environment instantiation; normalize actions using sigmoid

jamesheald committed:
organised and test environment registration; added target_jnt_range to pose env configs; general cleanup

jamesheald committed:
bug fix: target stacking; target dim calculation

jamesheald committed:
env make function added

jamesheald committed:
bug fix: data types

jamesheald committed:
minor modifications: np to jp; reshape reach_dist

Balint-H committed:
Add simplified learning script

CharlieLeee committed:
add mjx version installation guide

CharlieLeee committed:
remove extra tag in standard and mjx dependencies, update readme

CharlieLeee committed:
update readme

CharlieLeee committed:
update readme

CharlieLeee committed:
fix small mistake in env registration

jamesheald committed:
remove myosuite_scene to speed up simulation

jamesheald committed:
fix done bug

jamesheald committed:
add target sids for rendering targets

jamesheald committed:
use init qpos described in mujoco model, not random initialisations

jamesheald committed:
define max_episode_steps parameter; add step count; add reset logic in step function, to reset targets on termination or truncatio; add metric quantifying fraction of steps that are solved

jamesheald committed:
calculate n_substeps and pass to mjx_env.step to scan over multiple substeps with the same action

jamesheald committed:
use make_data instead of init to create data in reset

jamesheald committed:
create policy_parms_fn for rendering at evaluations throughout during and log videos and metrics to wandb

jamesheald committed:
use mujoco warp

jamesheald committed:
pose class modifications: add step count; add reset logic in step function; add fraction of steps solved metric; use mujoco warp; use make_data in reset; scan over multiple substeps with the same action in env.step

sagarverma committed:
jax utils and tests (#329)

Vittorio-Caggiano commented on myosuite/envs/myo/assets/elbow/myoelbow_1dof6muscles_mjx.xml:
can we use the sim_hive file? in case, can we generate it via mjspec?

Vittorio-Caggiano commented on myosuite/envs/myo/assets/elbow/myoelbow_1dof6muscles_mjx_eval.xml:
can we use the sim_hive file? in case, can we generate it via mjspec?

Balint-H replied:
We already do, these are no longer referenced in code! I removed them.

Vittorio-Caggiano commented on myosuite/envs/myo/assets/leg/myolegs_chasetag.xml:
can we use the sim_hive file? in case, can we generate it via mjspec?

Balint-H committed:
Remove hardcoded mjx environment files

Balint-H committed:
Mention warp in the README of the MJX repo

Balint-H reviewed (approved):
LGTM

CharlieLeee committed:
update dependency and installation guide (#355)

* update dependency and installation guide

* revert mujoco version

Vittorio-Caggiano committed:
add mjx unit test on CI

Vittorio-Caggiano committed:
add mjx test

Vittorio-Caggiano committed:
add pytest

Vittorio-Caggiano committed:
use python 3.10 for ci tests

CharlieLeee committed:
Fix NPZ format mismatch: convert 6D euler object_init to 7D quaternion format (#358)

Some NPZ files store object_init in 6D format (pos + euler angles) while
object trajectories use 7D format (pos + quaternion). This caused assertion
failures in test_load_npz_file.

Added automatic conversion in both NumPy and JAX implementations to detect
and convert 6D object_init to 7D quaternion format during load.

Vittorio-Caggiano committed:
remove dm_control (#361)

* remove dm_control
* add docstrings
Co-authored-by: Bálint Hodossy <bkh16@ic.ac.uk>

Vittorio-Caggiano committed:
Merge branch 'dev' into mjx

Vittorio-Caggiano committed:
run mjx ci test to uv

Vittorio-Caggiano committed:
explicitly tell `uv` which dependency extras to include

Vittorio-Caggiano committed:
remove tqdm

Vittorio-Caggiano committed:
remove ptr reference and pass directly the model

Vittorio-Caggiano committed:
fix dm_control removal

Vittorio-Caggiano committed:
Update docs and tutorials (#381)

Vittorio-Caggiano committed:
rebased

Vittorio-Caggiano committed:
fix docs

Vittorio-Caggiano committed:
fix ci on win

Vittorio-Caggiano committed:
skip doc compilation when no changes are made in the docs

jamesheald committed:
mjx-cuda extra dependencies

Vittorio-Caggiano committed:
test

Vittorio-Caggiano committed:
remove test file

jamesheald committed:
add imageio dependency to render videos

jamesheald committed:
change default impl from warp to jax

jamesheald committed:
bump MuJoCo and MJX and apply patch to use MJWarp via MJX API

jamesheald committed:
remove rendering during training option (slow - bad practice)

jamesheald committed:
warp as default impl as much faster

jamesheald committed:
benchmark results

jamesheald committed:
update README

jamesheald committed:
update README

jamesheald committed:
update figure

jamesheald committed:
add argparse to specify env_name and impl during training

fl0fischer committed:
MJX fixes + JAX Fatigue (#371)

* fixed nconmax and njmax args

* fixed missing/outdated dependencies

* bump to mujoco 3.4.0

* JaxFatigue: modified data structure to avoid side effects (return all dynamic values instead of storing them)

* register muscle variants along with the base env (only fatigue tested so far)

* add fatigue to reach envs and configs

* use data.userdata rather than info dict to store dynamic fatigue states

* move JAX fatigue script to mjx folder

* remove fatigue variants from mjx base classes

* rename key to rng for consistency; WARNING: rng is mandatory arg of CumulativeFatigue.reset()

* add FatigueWrapper class

* update fatigue config, fix variant registering, and optionally wrap FatigueWrapper when env is loaded

* minor code cleaning; fallback if obs keys change

* removed unused pytree utils functions

* fatigue obs: concatenate single axis

* code cleaning/restructuring

* Extract base class for MJX envs

* Fix unnecessary double function call

* Use more general method name

* Separate wrapper logic for config parsing

* Fix override passing in env registration

* Typo fix in readme

* fix fatigue tutorial

* replaced ffmpeg subprocess with imageio call in fatigue tutorial

* Fix video gitignore

* Add imageio dep

* benchmark results

* update README

* update README

* update figure

* add argparse to specify env_name and impl during training

* Adjust impl override in training script

* only disable cylinder contacts for jax impl

* fix issue with nconmax/naconmax; WARNING: requires bump to mujoco 3.5.1!

* Minor fixes in tutorials

* restructurings in MjxMyoBase; increase default constraint/optimization solver params (may result in 2-5% time increase, but required to run all benchmark tests without mujoco-mjx warnings); run step-count-increase logic before observations are made

* minor bugs fixed introduced when merging PR; remove unnecessary imports

* Delete =3.4.0

* Delete =

* Add normalization flag in mjx configs, adjust wording in tutorials

* Adjust observation naming

* Pin mujoco 3.5.0

---------

Co-authored-by: Balint-H <bkh16@ic.ac.uk>
Co-authored-by: jamesheald <jamesbheald@gmail.com>

Vittorio-Caggiano committed:
fix mujoco version

Vittorio-Caggiano committed:
bump mujoco version

Vittorio-Caggiano committed:
make mujoco version looser for now

Vittorio-Caggiano committed:
make mujoco a range

Vittorio-Caggiano committed:
remove git install from main mujoco

Vittorio-Caggiano committed:
add uv lock

fl0fischer committed:
increase default ccd_iterations and naconmax to avoid errors when training one of the three default envs

fl0fischer committed:
updated/extended benchmark results in mjx README

fl0fischer committed:
minor typo

fl0fischer committed:
added benchmark scripts

fl0fischer committed:
unnecessary import removed

Vittorio-Caggiano committed:
prevent dependency re-syncing during tests on macOS

Balint-H committed:
Add policy options to MJX visualisation script

Balint-H committed:
Add back linted imports to tutorial utils script

Vittorio-Caggiano committed:
fix unit test for ci

Vittorio-Caggiano committed:
remove macos from ci

Vittorio-Caggiano committed:
reintroduce macos

Vittorio-Caggiano committed:
fix tutorial and installation

Vittorio-Caggiano committed:
add video_path check for ci

Vittorio-Caggiano committed:
fix wrapper issue

Vittorio-Caggiano committed:
fix unwrapped env

Vittorio-Caggiano committed:
make safe writing for ci

Vittorio-Caggiano committed:
remove fatigue tutorial from CI

Vittorio-Caggiano committed:
remove fatigue tutorial from ci

fl0fischer committed:
tutorials fixed/updated; replaced skvideo with imageio; fixed tutorial dependencies

fl0fischer committed:
removed deprecated method 'examine_policy'

Vittorio-Caggiano committed:
add ffmpeg to imageio

Vittorio-Caggiano commented on .gitmodules:
```diff
@@ -1,6 +1,6 @@
 [submodule "myosuite/simhive/myo_sim"]
 	path = myosuite/simhive/myo_sim
-	url = https://github.com/MyoHub/myo_sim.git
+	url = https://github.com/Balint-H/myo_sim.git
```
@Balint-H is there any change you have locally that is needed in myo_sim or can we refer to the myo_sim repo on MyoHub?

Balint-H committed:
Update tutorial links in docs and deprl example

Vittorio-Caggiano committed:
add space in code-block

fl0fischer committed:
linked to benchmark scripts in MJX README

vikashplus committed:
BUGFIX: Remove state env registrations (#367)

Balint-H committed:
Updates submodule URL
